sdk/include/opentelemetry/sdk/resource/resource.h

```diff
@@ -0,0 +1,76 @@
+#pragma once
+
+#include "opentelemetry/sdk/common/attribute_utils.h"
+#include "opentelemetry/sdk/resource/resource_detector.h"
+#include "opentelemetry/sdk/version/version.h"
+#include "opentelemetry/version.h"
+
+#include <cstdlib>
+#include <memory>
+#include <sstream>
+#include <unordered_map>
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace resource
+{
+
+using ResourceAttributes =
+    std::unordered_map<std::string, opentelemetry::sdk::common::OwnedAttributeValue>;
+
+class Resource
+{
+public:
+  const ResourceAttributes &GetAttributes() const noexcept;
+
+  /**
+   * Returns a new, merged {@link Resource} by merging the current Resource
+   * with the other Resource. In case of a collision, current Resource takes
+   * precedence.
+   *
+   * @param other the Resource that will be merged with this.
+   * @returns the newly merged Resource.
+   */
+
+  std::shared_ptr<Resource> Merge(const Resource &other) noexcept;
```

Comment:
Do we have to implement these (internal to SDK surface) methods.. can we make it a `map` - inherited from standard library map (either `map` or `unordered_map`), and follow semantics of existing APIs available, e.g. [map::merge](https://en.cppreference.com/w/cpp/container/map/merge) .. Then we are not inventing some new Visual Basic-style API, but follow existing Standard Library semantics everywhere. This will be easier to comprehend by experienced modern C++ developer... Plus - I think it would be consistent if our container classes are compatible with existing `algorithm` templates [described here](https://en.cppreference.com/w/cpp/algorithm/merge)

Reply:
The spec defines [some strict rules](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/resource/sdk.md#merge) around merging, which I think are not satisfied by `map::merge`:

> * If the value on the primary resource is an empty string, the result has the value of the secondary resource.
> * Otherwise, the value of the primary resource is used.